src/mem/pagemap.h

```diff
@@ -381,7 +379,15 @@ namespace snmalloc
      */
     size_t index_for_address(uintptr_t p)
     {
-      return reinterpret_cast<size_t>((p >> SHIFT) & (OS_PAGE_SIZE - 1));
+      size_t paddr;
+
+#if defined(__CHERI_PURE_CAPABILITY__)
+      paddr = static_cast<size_t>(static_cast<vaddr_t>(p));
```

Comment:
Is this incorrect in the non-CHERI case?  It seems that this should work both ways.